volatility3/framework/symbols/linux/extensions/__init__.py

```diff
@@ -1761,3 +1764,132 @@ def get_capabilities(self) -> int:
             )
 
         return cap_value & self.get_kernel_cap_full()
+
+
+class timespec64(objects.StructType):
+    def to_datetime(self) -> datetime:
+        """Returns the respective aware datetime"""
+
+        dt = renderers.conversion.unixtime_to_datetime(self.tv_sec + self.tv_nsec / 1e9)
+        return dt
+
+
+class inode(objects.StructType):
+    def is_valid(self) -> bool:
+        # i_count is a 'signed' counter (atomic_t). Smear, or essentially a wrong inode
+        # pointer, will easily cause an integer overflow here.
+        return self.i_ino > 0 and self.i_count.counter >= 0
+
+    def is_dir(self) -> bool:
+        """Returns True if the inode is a directory"""
+        return stat.S_ISDIR(self.i_mode) != 0
+
+    def is_reg(self) -> bool:
+        """Returns True if the inode is a regular file"""
+        return stat.S_ISREG(self.i_mode) != 0
+
+    def is_link(self) -> bool:
+        """Returns True if the inode is a symlink"""
+        return stat.S_ISLNK(self.i_mode) != 0
+
+    def is_fifo(self) -> bool:
+        """Returns True if the inode is a FIFO"""
+        return stat.S_ISFIFO(self.i_mode) != 0
+
+    def is_sock(self) -> bool:
+        """Returns True if the inode is a socket"""
+        return stat.S_ISSOCK(self.i_mode) != 0
+
+    def is_block(self) -> bool:
+        """Returns True if the inode is a block device"""
+        return stat.S_ISBLK(self.i_mode) != 0
+
+    def is_char(self) -> bool:
+        """Returns True if the inode is a char device"""
+        return stat.S_ISCHR(self.i_mode) != 0
+
+    def is_sticky(self) -> bool:
+        """Returns True if the sticky bit is set"""
+        return (self.i_mode & stat.S_ISVTX) != 0
+
+    def get_inode_type(self) -> str:
+        """Returns inode type name
+
+        Returns:
+            The inode type name
+        """
+        if self.is_dir():
+            return "DIR"
+        elif self.is_reg():
+            return "REG"
+        elif self.is_link():
+            return "LNK"
+        elif self.is_fifo():
+            return "FIFO"
+        elif self.is_sock():
+            return "SOCK"
+        elif self.is_char():
+            return "CHR"
+        elif self.is_block():
+            return "BLK"
+        else:
+            return renderers.UnparsableValue()
+
+    def get_inode_number(self) -> int:
+        """Returns the inode number"""
+        return int(self.i_ino)
```

Comment:
Is this not already an int?  Just wondering why the overly cautious casting?  Is it so it can go straight into the results (a volatility int should still be ok there)?  I prefer not casting unnecessarily because other people see it and then they can unnecessarily because they think they have to, and soon everything is littered with unnecessary casts and it becomes much more difficult to read.

Reply:
Ok, I will change this

Reply:
The problem with returning non-standard types for things like simple numbers is that it messes up serialization and makes deepcopying really really slow. It's hard to spot and fix unless you're a pretty advanced Python developer. IMO it's not good idea to do this but up to you.

Reply:
I try to ensure that all parameters are standard types rather than complex objects for similar thinking, but the point is people should be able to treat this as an int and not care (because its base class *is* `int`).  As to deepcopying, that's usually not a good solution, but one that was necessary when it was used, so I don't see it as a huge reason.  Serialization (particularly if we ever want to parallelize this somehow) is more important but again, the property is getting called, so the value isn't getting stashed and the receiving party can decide what to do with it, but I've seen a lot of example code littered with these stray casts, not because they were concerned about performance, but because they didn't know whether they could just use it or not.  As such, I try hard to weed them out, so it's probably best to be consistent.  If there's an actual use case where it causes a massive speed difference I'll definitely consider it, but without that it should just be returned (although really, at this point, they're just acting as variable renames)?

Reply:
Yeah, that's the point. Without the cast that method works only as a more informative/clear name to get the inode number. The whole point of it was to make sure, with the function name and with the typing, that the user wants the inode integer number. At the end of the day, if the user wants the original type for any reason, this doesn't prevent the user of directly calling `inode.i_ino` instead and obtain the objects.Integer.

> people should be able to treat this as an int and not care (because its base class is int).

Hm, inheriting from `int` doesn't necessarily mean you can use them interchangeably everywhere. `int` is an atomic, immutable type, while `objects.Integer` is not. Of course, for basic stuff like doing math operations, comparisons and so it will work like an `int`.
For instance, you will have issues using this type with the standard [dataclasses](https://docs.python.org/3/library/dataclasses.html) when calling [astuple](https://docs.python.org/3/library/dataclasses.html#dataclasses.astuple). It will need to do a deepcopy taking a long time doing that conversion.

Anyway, back to `get_inode_number()` ... Do you still find it odd to have a method that returns an int() cast like that?

Reply:
So possibly I'm a little disappointed that I put all this work into making sure people could use them exactly like ints (because they are), and then users still don't know/trust that it's the case and end up littering the code with `int` or losing valuable information about where the object came from because they threw it away without good reason.

I tested it, you're right, it does take a long time doing `astuple`, but it does complete just fine.  I think it should be the consumer that optimizes rather than producer, because it only needs optimizing for certain very specific use cases and otherwise the origin information may be useful.

Volshell snippet for a linux image:
```
>>> import dataclasses
>>> @dataclasses.dataclass
... class Point:
...      x: int
...      y: int

>>> volpoint = Point(ps()[0].pid, ps()[1].pid)
>>> dataclasses.astuple(volpoint)
(1, 2)
```

So yeah, I still find it a bit odd using it just to cast, and whilst you say people can just go back to the original, if all our examples and code only use the properties, they likely won't go digging to find out the right name for something that's already supposed to be exposed.  If we're putting all those properties in place just to wrap them in `int`, on the off-chance they may one day get deep-copied, I don't think that's a strong enough justification and I'd prefer we just used the original names and scrap those properties (unless they're overly cumbersome to type/unintuitive), because staying close to the source ensures that when people discuss stuff they're not using volatility-decided pseudonyms, they can discuss it with kernel developers and be using the actual struct members so everyone's on the same page.

I think adding this will just muddy people's idea of what goes on, degrade the consistency of the framework and I think it will just ultimately lead to more people wrapping things as ints without knowing why and further diluting the object model.  Having said all that I don't want to fight about this too much, and my will to keep arguing it is diminishing so if you really think it's vital that we provide an int wrapped convenience function because it's a significant issue with the core object model of the framework which will impact a lot of people very often, then I'll relent.

Reply:
I think it may actually just be this one property, for some reason I thought it was a whole slew of them...  That's less trend setting, so I'm not quite as upset letting it slip through, in general though, I tell people to ditch `int` wrappers unless they can come up with a really good reason (and too often when I ask it's been "oh I just thought these wouldn't work" or "oh I copied it from vol 2 where you had to and thought it would be the same").

Reply:
Fair enough, I agree that keeping the original names is a good practice. Plus, it's not a big deal to leave it as is and handle the casting on the consumer side. It was good anyway having this chance to discuss and be aware of some side-effects of not using the built-in types. Thanks

Reply:
> I think it may actually just be this one property, for some reason I thought it was a whole slew of them... That's less trend setting, so I'm not quite as upset letting it slip through, in general though, I tell people to ditch `int` wrappers unless they can come up with a really good reason (and too often when I ask it's been "oh I just thought these wouldn't work" or "oh I copied it from vol 2 where you had to and thought it would be the same").

Ups, for some reason I missed your last comment when I replied above. 

You first comment was already quite convincing to me :) instead, it seems we convinced each other simultaneously :) LOL

> I think it may actually just be this one **property**

Sorry, I didn't catch that. Currently, the function isn't a property. When you say `property`, do you mean we should convert it to one, or are you just referring to the function?
Do you think it's better to keep the member original name (removing entirely this function) or;  returning an int() with either a function, or a property?

Reply:
Sorry, it starts with get so it can stay as a function (although that seems less convenient to me).  If we convinced each other, then I'd much prefer to lean towards ditching the int cast (and if necessary the whole function) please.

Reply:
Great, here we go! Hopefully, I didn't mess anything up. Let's see what the test cases say